src/lfu.ml

```diff
@@ -0,0 +1,101 @@
+module Make (K : sig
+  type t
+
+  val equal : t -> t -> bool
+  val hash : t -> int
+end) =
+struct
+  module H = Hashtbl.Make (K)
+
+  type key = K.t
+
+  type 'a t = {
+    value : (int * int * 'a) H.t;
+    frequency : (int, key Dllist.t) Hashtbl.t;
+    mutable size : int;
+    cap : int;
+    stats : Stats.t;
+  }
+
+  (*let dummy : K.t = Obj.magic (ref 0)*)
+
+  let unsafe_v c =
+    {
+      value = H.create c;
+      frequency = Hashtbl.create c;
+      size = 0;
+      cap = c;
+      stats = Stats.v ();
+    }
+
+  let v c =
+    if c <= 0 then invalid_arg "capacity must be strictly positive";
+    unsafe_v c
+
+  let stats t = t.stats
+  let is_empty t = H.length t.value = 0
+  let capacity t = t.cap
+  let size t = H.length t.value
+
+  let clear t =
+    H.clear t.value;
+    Hashtbl.clear t.frequency;
+    Stats.clear t.stats
+
+  let mem t k =
+    let b = H.mem t.value k in
+    if b then Stats.hit t.stats else Stats.miss t.stats;
+    b
+
+  let hidden_find t k =
+    let freq, index, value = H.find t.value k in
+    let dl = Hashtbl.find t.frequency freq in
+    (freq, index, dl, value)
+
+  let find t k =
+    let freq, index, odl, value = hidden_find t k in
+    let ndl = Hashtbl.find t.frequency (freq + 1) in
+    Dllist.remove odl index;
+    let index, removed = Dllist.append ndl k in
```

Comment:
Instead of exposing these integers indices that make only sense to the `Dllist` abstraction, I think it would be better to make these abstract and call them something like `dllist_key`.

@pascutto WDYT